Classes/Http/RedirectMiddleware.php

```diff
@@ -0,0 +1,95 @@
+<?php
+declare(strict_types=1);
+namespace Yeebase\TwoFactorAuthentication\Http;
+
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\Mvc\ActionRequest;
+use Neos\Flow\Mvc\Routing\UriBuilder;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use GuzzleHttp\Psr7\Response;
+
+/**
+ * A HTTP component that redirects to the configured 2FA login/setup routes if requested
+ */
+final class RedirectMiddleware implements MiddlewareInterface
+{
+    public const REDIRECT_LOGIN = 'login';
+    public const REDIRECT_SETUP = 'setup';
+
+    /**
+     * @Flow\InjectConfiguration(path="routes.login")
+     * @var array
+     */
+    protected $loginRouteValues;
+
+    /**
+     * @Flow\InjectConfiguration(path="routes.setup")
+     * @var array
+     */
+    protected $setupRouteValue;
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $next): ResponseInterface
+    {
+        $response = $next->handle($request);
+        // TODO: Find a better solution to communicate from TokenProvider to this middleware
+        $redirectTarget = $response->getHeaderLine(static::class . '.redirect');
```

Comment:
See https://github.com/neos/flow-development-collection/issues/2019#issuecomment-720463971

Though we do not even have access to the `ActionResponse` in the TokenProvider, so this might become a bit harder to achieve than anticipated. If anyone has a good idea, just shoot.
I'm still thinking if it wouldn't make more sense to use the `entryPoint` configuration for the login redirect and use a middleware that checks for the setup redirect.

Reply:
Can't this be solved with exceptions, as is the case here?

https://github.com/sandstorm/NeosTwoFactorAuthentication/blob/main/Classes/Security/Authentication/Provider/PersistentUsernameAndPasswordWithSecondFactorProvider.php#L68

Reply:
using exceptions for control flow is a bit quirky because of the overhead of throwing/catching exceptions (i.e. they have to create a full stack trace).
But in this case I also think that it's the nicer solution – and one top-level exception shouldn't do any harm

Reply:
@gerdemann btw I think your solution could be simplified to
```php
    public function process(ServerRequestInterface $request, RequestHandlerInterface $next): ResponseInterface
    {
        try {
            $response = $next->handle($request);
        } catch (\Exception $exception) {
            if ($exception instanceof SecondFactorLoginException || $exception->getPrevious() instanceof SecondFactorLoginException) {
                 return $this->redirectToLogin($request);
            } elseif ($exception instanceof SecondFactorSetupException || $exception->getPrevious() instanceof SecondFactorSetupException) {
                 return $this->redirectToSetup($request);
            } else {
                throw $exception;
            }
        }
        return $response;
    }
```

Reply:
Thank you, I have now adjusted this and created a PR. It doesn't matter to me whether this PR is adjusted or mine is merged. 😃 
https://github.com/yeebase/Yeebase.TwoFactorAuthentication/pull/8